meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClient.java

```diff
@@ -115,6 +116,52 @@ public void create(String key, Object data, MetaClientInterface.EntryMode mode)
     }
   }
 
+  @Override
+  public void recursiveCreate(String key, T data, EntryMode mode) {
+    // Function named recursiveCreate to match naming scheme, but actual work is iterative
+    iterativeCreate(key, data, mode, -1);
+  }
+
+  @Override
+  public void recursiveCreateWithTTL(String key, T data, long ttl) {
+    iterativeCreate(key, data, EntryMode.TTL, ttl);
+  }
+
+  private void iterativeCreate(String key, T data, EntryMode mode, long ttl) {
+    List<String> nodePaths = separateIntoUniqueNodePaths(key);
+    int i = 0;
+    // Ephemeral nodes cant have children, so change mode when creating parents
+    EntryMode parentMode = (EntryMode.EPHEMERAL.equals(mode) ?
+        EntryMode.PERSISTENT : mode);
+
+    // Iterate over paths, starting with full key then attempting each successive parent
+    // Try /a/b/c, if fails due to NoNode then try /a/b .. etc..
+    while (i < nodePaths.size()) {
+      try {
+        if (EntryMode.TTL.equals(mode)) {
+          createWithTTL(nodePaths.get(i), data, ttl);
+        } else {
+          create(nodePaths.get(i), data, i == 0 ? mode : parentMode);
+        }
+        break;
+        // NoNodeException thrown when parent path does not exist. We allow this and re-attempt
+        // creation of these nodes below
+      } catch (MetaClientNoNodeException ignoredParentDoesntExistException) {
+        i++;
+      }
+
+    }
+
+    // Reattempt creation of children that failed due to NoNodeException
+    while (--i >= 0) {
+      if (EntryMode.TTL.equals(mode)) {
+        createWithTTL(nodePaths.get(i), data, ttl);
+      } else {
+        create(nodePaths.get(i), data, i == 0 ? mode : parentMode);
+      }
+    }
+  }
```

Comment:
This require try catch that if two threads parallel creation the children at the same time. Also I would imagine this should be same path as previous one.

@xyuanlu could you please help @GrantPSpencer walk through the idea?

Reply:
I'm not sure I fully understand this race condition yet. I'll sync up with @xyuanlu to get more info

Reply:
Changed to ignore when NodeExistsException thrown when we try to create the parents, only throw when we try to create the full node path. This means we don't throw an error if another thread has already created the parents we are trying to create.